### PR TITLE
318 multi scope search

### PR DESCRIPTION
**Pull request by clarkepeterf**

### CODE REVIEW

**Comment by brent-hartwig** [comment]
@clarkepeterf and @roamye,

Please discuss and decide if we want to make changes now or see if it becomes a problem.

When one of the OR'd items is rejected, the error message does not explicitly identify which one.  Further, at least one error condition could be conditionally ignored: [only-ignored-terms](https://github.com/project-lux/lux-marklogic/pull/370/files#diff-3c0e47d41537ce3eb7cd3dd213fffe49ddeb8622c5aa3ab7894ee64789942edaR180-R194).  Looking over the other `InvalidSearchRequestError` instances, that one might be the most commonly encountered by users.  

The following search is an example of the above.  It errors out with:

> Invalid search request: the search criteria given only contains 'the', which is an ignored term(s). Please consider creating phrases using double quotes and/or adding additional criteria."

If the system ignored the second one, it could have returned results without changing the meaning of the search.

But, if this is indeed the most likely error condition, the error message could be sufficient (even if not correctly worded for a multi scope search).

```json
{
  "_scope":"multi",
  "OR":[
    {
      "_scope":"item",
      "text":"chair"
    },
    {
      "_scope":"work",
      "text":"the"
    }
  ]
}
```

**Comment by roamye** [comment]
@clarkepeterf - should this be discussed at IT team meeting?

File: src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs [hunk 1]
```diff
@@ -191,6 +242,67 @@ const SearchCriteriaProcessor = class {
...
+    const docPlan = op
+      .fromSearch(ctsQuery)
+      .joinDocUri('uri', op.fragmentIdCol('fragmentId'));
+
+    const { order, subSortConfigs } =
+      this.sortCriteria.getMultiScopeSortOption();
+    const subSortPlans = subSortConfigs.map((subSortConfig) =>
+      this._getSubSortPlan(docPlan, subSortConfig)
+    );
+    const unionPlan = subSortPlans.reduce((union, curr) => {
```
[the remark this answers is not in the document]
Comment (brent-hartwig) [reply src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs:266]: @clarkepeterf, please check out [318-pipeline.js.txt](https://github.com/user-attachments/files/17706877/318-pipeline.js.txt).  Primary differences:

1. Switches from `.union` to `.joinFullOuter`, as described above.
2. Calls `.fromSearch` once.
3. Calls `.fromLexicons` once.
4. Joins those two on the fragment ID (concurrency advantage).
5. Drops my first suggestion's use of `.groupBy` as the duplicate rows were due to the second `.fromSearch`.
6. Defers the URI join until after pagination.

[the remark this answers is not in the document]
Comment (brent-hartwig) [reply src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs:266]: BTW, `.export` helped me here.  In `_getMultiScopeSortResults`, I set a new variable to the plan, up to calling `.result`, logged it via `console.log(JSON.stringify(finalPlan.export()))`, removed the return characters, pasted it into `op.toSource()` in QC, then had to code to work over.

File: src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs [hunk 2]
```diff
@@ -118,6 +115,60 @@ const SearchCriteriaProcessor = class {
...
+              searchCriteriaProcessor.process(
+                subCriteria,
+                scopeName,
+                searchPatternOptions,
+                includeTypeConstraint,
+                page,
+                pageLength,
+                sortCriteria,
+                valuesOnly
+              );
```
Comment (brent-hartwig) [inline src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs:157]: I'm not sure a comma needs to be added.  Seems to be working with or without.

File: src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs [hunk 3]
```diff
@@ -86,11 +87,7 @@ const SearchCriteriaProcessor = class {
     sortCriteria,
     valuesOnly
   ) {
```
Comment (brent-hartwig) [inline src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs:89]: @clarkepeterf, I'm working up a variation of this that keeps `_requireSearchCriteriaJson` in SearchCriteriaProcessor.  Moving it to searchLib's `processSearchCriteria` is presently safe but should we ever introduce a second call to `process`, we'd have to duplicate the check.  I've run out of time for the day but plan on testing my edits tomorrow morning.

**Comment by clarkepeterf** [comment]
@brent-hartwig  @roamye  I've created added a commit which makes the errors for subCriteria specific about which scope is causing the error: https://github.com/project-lux/lux-marklogic/pull/370/commits/f44e296405d99563b6591774400d0ce992cd49d1

This addresses [this comment above](https://github.com/project-lux/lux-marklogic/pull/370#issuecomment-2468507711):
>@clarkepeterf and @roamye,
>
>Please discuss and decide if we want to make changes now or see if it becomes a problem.
>
>When one of the OR'd items is rejected, the error message does not explicitly identify which one. Further, at least one error condition could be conditionally ignored: [only-ignored-terms](https://github.com/project-lux/lux-marklogic/pull/370/files#diff-3c0e47d41537ce3eb7cd3dd213fffe49ddeb8622c5aa3ab7894ee64789942edaR180-R194). Looking over the other InvalidSearchRequestError instances, that one might be the most commonly encountered by users.
>...

File: src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (clarkepeterf) [reply src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs:272]: Good point, I was mostly working with the archives that will be using this search, so they all have values in those indexes. Will test out your updated code

**Comment by brent-hartwig** [comment]
> @brent-hartwig @roamye I've created added a commit which makes the errors for subCriteria specific about which scope is causing the error: [f44e296](https://github.com/project-lux/lux-marklogic/commit/f44e296405d99563b6591774400d0ce992cd49d1)

Thanks, Peter.  That outta help.  If the issue is in the 'item' scope, that's the scope name that will be included in the message to the user, as opposed to 'object'; however, that's already possible in other messages.

File: src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs (hunk 3 above)
Comment (brent-hartwig) [reply src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs]: @clarkepeterf, I concluded `requireSearchCriteriaJson` was moved to accommodate the new nested calls of `process` but found another way: those nested calls can just pass `null` in for the scope name, thereby requiring the scope from the search criteria's `_scope` property (https://github.com/project-lux/lux-marklogic/pull/370/commits/325211b2047b08f2e30b9c1c257ba58aae1180da).  Please advise if I'm creating a problem 😜

**Comment by brent-hartwig** [comment]
@clarkepeterf, https://github.com/project-lux/lux-marklogic/pull/370/commits/b2672ad66c5a45af34a53ea61774343fc2a50c2a blocks the user of the multi search scope in facet requests, related list requests, and on individual items within a multi search scope request's OR array.  We don't have to be this strict but but for facets in particular, I believe:

1. The facets endpoint consumer should be responsible for only specifying legitimate criteria.
2. The facets endpoint should trim the criteria to the legitimate criteria.
3. The facets endpoint should reject the request.

I went with no. 3 as nested multi search scope requests also came to mind.  I defaulted to us not needing or wanting to support that at this time.  Thoughts?

File: src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs (hunk 3 above)
Comment (clarkepeterf) [reply src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs]: Looks good to me

File: src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs [hunk 4]
```diff
@@ -117,6 +119,67 @@ const SearchCriteriaProcessor = class {
...
+                  includeTypeConstraint,
+                  page,
+                  pageLength,
+                  sortCriteria,
+                  valuesOnly
+                );
+              } catch (e) {
+                e.message = `Error in scope '${subScope}': ${e.message}`;
+                throw e;
+              }
```
[the remark this answers is not in the document]
Comment (clarkepeterf) [reply src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs:179]: Noted and confirmed via testing on my side. Removing the comma in my next commit.

File: src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (clarkepeterf) [reply src/main/ml-modules/root/lib/SearchCriteriaProcessor.mjs]: New code returns values that don't match the sorting indexes

**Pull request merged**